### PR TITLE
Constrained search

### PR DESCRIPTION
**Pull request by jdickerson95**
Adds a program for doing constrained searches—also a few minor changes to the core to add functionality/improve flexibility.

Tests are failing because it's using the torch-so3 updates that have not been released yet.

It's finished,d subject to cleaning a few things up and documentation. Needs some more testing before merging as well.

### CODE REVIEW

File: src/leopard_em/pydantic_models/orientation_search.py [hunk 1]
```diff
@@ -146,23 +178,105 @@ def euler_angles_offsets(self) -> torch.Tensor:
...
+    psi_min: float
+        Minimum value for the psi angle in degrees.
+    psi_max: float
+        Maximum value for the psi angle in degrees.
+    """
+
+    enabled: bool = True
+    in_plane_step: float = 1.5
+    out_of_plane_step: float = 2.5
+    rotation_axis_euler_angles: list[float] = Field(default=[0.0, 0.0, 0.0])
```
Comment (mgiammar) [inline src/leopard_em/pydantic_models/orientation_search.py:222]: This should only permit one rotation axis (three floats), correct? Then this should be typehinted as a `tuple[float, float, float]` to ensure exactly three Euler angles are passed to this attribute field.

File: src/leopard_em/pydantic_models/orientation_search.py [hunk 2]
```diff
@@ -146,23 +178,105 @@ def euler_angles_offsets(self) -> torch.Tensor:
...
+    phi_min: float = 0.0
+    phi_max: float = 360.0
+    theta_min: float = 0.0
+    theta_max: float = 180.0
+    psi_min: float = 0.0
+    psi_max: float = 360.0
+    base_grid_method: Literal["uniform", "healpix", "basic"] = "uniform"
+
+    @property
+    def euler_angles_offsets(self) -> torch.Tensor:
```
Comment (mgiammar) [inline src/leopard_em/pydantic_models/orientation_search.py:232]: Update return type to `tuple[torch.Tensor, torch.Tensor]` and the docstring accordingly for the offsets and rotated Euler angles

File: src/leopard_em/pydantic_models/particle_stack.py [hunk 3]
```diff
@@ -165,8 +165,20 @@ def construct_image_stack(
             img = load_mrc_image(img_path)
 
             # with reference to center pixel
-            pos_y = self._df.loc[indexes, "pos_y_img"].to_numpy()
-            pos_x = self._df.loc[indexes, "pos_x_img"].to_numpy()
+            pos_y = self._df.loc[
```
Comment (mgiammar) [inline src/leopard_em/pydantic_models/particle_stack.py:168]: Would prefer the if/else statement was brought outside of the dataframe indexing. Also should update docstring accordingly to say this method preferentially selects refined columns if they are present

File: src/programs/constrained_search/constrained_search_example_config.yaml [hunk 4]
```diff
@@ -0,0 +1,43 @@
+###################################################
```
Comment (mgiammar) [inline src/programs/constrained_search/constrained_search_example_config.yaml:1]: Alignment of the octothorp symbols in the header

File: src/programs/constrained_search/utils/get_centre_vector.py
Comment (mgiammar) [inline src/programs/constrained_search/utils/get_centre_vector.py:1]: I see we are mixing British and American english spellings for file names xD

File: src/programs/constrained_search/utils/get_centre_vector.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
+"""Calculates the centre vector between two PDB structures."""
+
+import argparse
+import math
```
Comment (mgiammar) [inline src/programs/constrained_search/utils/get_centre_vector.py:4]: Use `np.pi` over `math.pi` to reduce the number of imports. I don't see the math module being used for anything but the constant

File: src/programs/constrained_search/utils/get_centre_vector.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    theta = torch.acos(norm_vector[2])
+
+    # Calculate phi as the azimuthal angle in the x-y plane
+    phi = torch.atan2(norm_vector[1], norm_vector[0])
+
+    # For a direction vector, psi can be set to 0
+    psi = torch.tensor(0.0)
+
+    # Convert to degrees
+    phi_deg = phi.item() * 180 / math.pi
```
Comment (mgiammar) [inline src/programs/constrained_search/utils/get_centre_vector.py:88]: Actually, instead of using the `math.pi` constant or `np.pi` constant, use the built-in `np.rad2deg` function for readability.

**Comment by mgiammar** [comment]
Should be ready for the new torch-so3 release for the phi/theta/psi step arguments. Also, the pylint is now failing on some of the new constrained orientation search portions of the code, but these mostly seem related to duplicate code for the refine template, etc.

File: programs/constrained_search/utils/get_rot_axis.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+sys.argv = [
+    "rotate_pdbs.py",
+    "3j77_aligned_SSU.pdb",
+    "3j78_aligned_SSU.pdb",
+    "rotation_axis.txt",
+]
+
+
+def calculate_rotation_matrix(
```
Comment (mgiammar) [inline programs/constrained_search/utils/get_rot_axis.py:17]: https://naver.github.io/roma/#roma.utils.rigid_vectors_registration

https://naver.github.io/roma/#roma.utils.rigid_points_registration

File: programs/constrained_search/utils/get_rot_axis.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+    # Handle reflection case
+    det = torch.det(R)
+    if det < 0:
+        Vt[-1, :] *= -1
+        R = Vt.T @ U.T
+
+    return R
+
+
+def extract_rotation_axis_angle(R: torch.Tensor) -> tuple[torch.Tensor, float]:
```
Comment (mgiammar) [inline programs/constrained_search/utils/get_rot_axis.py:52]: Naming convention argument should be rotmat

File: programs/constrained_search/utils/max_min_offsets.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+            print()
+
+    except FileNotFoundError:
+        print(f"Error: File '{csv_file}' not found.")
+    except pd.errors.EmptyDataError:
+        print(f"Error: '{csv_file}' is empty.")
+    except pd.errors.ParserError:
+        print(f"Error: '{csv_file}' is not a valid CSV file.")
+    except Exception as e:
+        print(f"Error: {e!s}")
```
Comment (mgiammar) [inline programs/constrained_search/utils/max_min_offsets.py:66]: Why is this wrapped in a try/except block? These errors would be raised by Pandas anyway

File: refine_template_example_config.yaml [hunk 10]
```diff
@@ -15,10 +15,10 @@ defocus_refinement_config:
...
 orientation_refinement_config:
   enabled: true
-  in_plane_angular_step_coarse:     1.5   # in degrees
-  in_plane_angular_step_fine:       0.15  # in degrees
-  out_of_plane_angular_step_coarse: 2.5   # in degrees
-  out_of_plane_angular_step_fine:   0.25  # in degrees
+  psi_step_coarse:     1.5   # in degrees
+  psi_step_fine:       0.15  # in degrees
+  theta_step_coarse: 2.5   # in degrees
+  theta_step_fine:   0.25  # in degrees
```
Comment (mgiammar) [inline refine_template_example_config.yaml:21]: Align values to the left

File: src/leopard_em/pydantic_models/config/orientation_search.py [hunk 11]
```diff
@@ -34,35 +40,35 @@ class OrientationSearchConfig(BaseModel2DTM):
...
-    phi_max: float = 360.0
-    base_grid_method: Literal["uniform", "healpix"] = "uniform"
+    psi_step: Annotated[float, Field(ge=0.0)] = 1.5
+    theta_step: Annotated[float, Field(ge=0.0)] = 2.5
+    phi_min: float | None = None
+    phi_max: float | None = None
+    theta_min: float | None = None
+    theta_max: float | None = None
+    psi_min: float | None = None
+    psi_max: float | None = None
```
Comment (mgiammar) [inline src/leopard_em/pydantic_models/config/orientation_search.py:70]: Change to typehint `Optional[float]`

File: src/leopard_em/pydantic_models/managers/constrained_search_manager.py [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+            self.centre_vector = torch.tensor(self.centre_vector, dtype=torch.float32)
+        rotation_matrices = roma.rotvec_to_rotmat(
+            roma.euler_to_rotvec(convention="ZYZ", angles=euler_angles)
+        ).to(torch.float32)
+        rotated_vectors = rotation_matrices @ self.centre_vector
+
+        # Get z-component for each particle individually
+        new_z_diffs = rotated_vectors[
+            :, 2
+        ]  # This is now a tensor with shape [batch_size]
```
Comment (mgiammar) [inline src/leopard_em/pydantic_models/managers/constrained_search_manager.py:130]: Move comment up a line to increase interpretability

File: src/leopard_em/pydantic_models/managers/constrained_search_manager.py [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+    #     results_csv_path : str
+    #         Path to the CSV file containing the per-particle data.
+
+    #     Returns
+    #     -------
+    #     RefineTemplateManager
+    #     """
+    #     df = pd.read_csv(results_csv_path)
+
+    #     return cls.from_dataframe(df)
```
Comment (mgiammar) [inline src/leopard_em/pydantic_models/managers/constrained_search_manager.py:403]: Remove unused commented out lines

**Review by mgiammar — CHANGES_REQUESTED** [review]

File: constrained_search_example_config.yaml [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+# Call `RefineTemplateManager.from_yaml(path)` to load this configuration.
+template_volume_path: /some/path/to/template.mrc # Volume of small particle
+particle_stack_large: # This is from the large particles
+  df_path: /some/path/to/particles.csv  # Needs to be readable by pandas
+  extracted_box_size: [528, 528]
+  original_template_size: [512, 512]
+particle_stack_small: # This is from the small particles
+  df_path: /some/path/to/particles.csv  # Needs to be readable by pandas
+  extracted_box_size: [528, 528]
+  original_template_size: [512, 512]
```
[the remark this answers is not in the document]
Comment (jdickerson95) [reply constrained_search_example_config.yaml:14]: Changed to reference_particle and constrained_particle

File: src/leopard_em/pydantic_models/managers/constrained_search_manager.py [hunk 15]
```diff
@@ -0,0 +1,370 @@
+"""Pydantic model for running the refine template program."""
```
Comment (mgiammar) [inline src/leopard_em/pydantic_models/managers/constrained_search_manager.py:1]: Running the constrained search program

File: src/leopard_em/pydantic_models/managers/refine_template_manager.py [hunk 16]
```diff
@@ -64,6 +63,8 @@ class RefineTemplateManager(BaseModel2DTM):
     preprocessing_filters: PreprocessingFilters
     computational_config: ComputationalConfig
 
+    snr_threshold: float = 1.0
```
Comment (mgiammar) [inline src/leopard_em/pydantic_models/managers/refine_template_manager.py:66]: Remove unused
